src/compiler/utilities.ts

```diff
@@ -10761,3 +10779,578 @@ export function createEvaluator({ evaluateElementAccessExpression, evaluateEntit
+                        }
+                    }
+                    break;
+                case SyntaxKind.BindingElement:
+                    if (
+                        lastLocation && (
+                            lastLocation === (location as BindingElement).initializer ||
+                            lastLocation === (location as BindingElement).name && isBindingPattern(lastLocation)
+                        )
+                    ) {
+                        if (isParameterDeclaration(location as BindingElement) && !associatedDeclarationForContainingInitializerOrBindingName) {
+                            associatedDeclarationForContainingInitializerOrBindingName = location as BindingElement;
+                        }
+                    }
+                    break;
+                case SyntaxKind.InferType:
+                    if (meaning & SymbolFlags.TypeParameter) {
+                        const parameterName = (location as InferTypeNode).typeParameter.name;
+                        if (parameterName && name === parameterName.escapedText) {
+                            result = (location as InferTypeNode).typeParameter.symbol;
+                            break loop;
+                        }
+                    }
+                    break;
+                case SyntaxKind.ExportSpecifier:
+                    // External module export bindings shouldn't be resolved to local symbols.
+                    if (
+                        lastLocation &&
+                        lastLocation === (location as ExportSpecifier).propertyName &&
+                        (location as ExportSpecifier).parent.parent.moduleSpecifier
+                    ) {
+                        location = location.parent.parent.parent;
+                    }
+                    break;
+            }
+            if (isSelfReferenceLocation(location)) {
+                lastSelfReferenceLocation = location;
+            }
+            lastLocation = location;
+            location = isJSDocTemplateTag(location) ? getEffectiveContainerForJSDocTemplateTag(location) || location.parent :
+                isJSDocParameterTag(location) || isJSDocReturnTag(location) ? getHostSignatureFromJSDoc(location) || location.parent :
+                location.parent;
+        }
+
+        // We just climbed up parents looking for the name, meaning that we started in a descendant node of `lastLocation`.
+        // If `result === lastSelfReferenceLocation.symbol`, that means that we are somewhere inside `lastSelfReferenceLocation` looking up a name, and resolving to `lastLocation` itself.
+        // That means that this is a self-reference of `lastLocation`, and shouldn't count this when considering whether `lastLocation` is used.
+        if (isUse && result && (!lastSelfReferenceLocation || result !== lastSelfReferenceLocation.symbol)) {
+            result.isReferenced! |= meaning;

```

Comment:
How safe is logic like this to exist outside of the checker/binder?

Reply:
Oh, this isn't safe *at all* in the first place. It's mutating a field directly on the `Symbol`, which will never be reset in incremental compilation in the language service, but whose value *should* change as the file does. I don't know how this crops up as a visible error in the language service, but it almost definitely does. But it also *has* for... forever? And we haven't run into it yet? So.... pass. Deal with it later?

Reply:
Of the 3 kinds of reference tracking we doing, this one is like number 1 most likely to be buggy, but it's only used for unused reference tracking, unlike `NodeLinks["isVisible"]` and `NodeLinks["referenced"]` both of which specifically track usages for imports in `.d.ts` and `.js` emit, respectively. So bugs in this one are probably rare to see reported, since usually they only result in missing suggestion diagnostics.

Reply:
...But if you do happen to know of issues along the lines of "this isn't marked as unused after I delete the code using it in this way" it's because this field here isn't on `SymbolLinks`.

Reply:
I'm going to guess that it doesn't crop up as a problem because within a file that changes, we always blow away the information; in a file that stays the same, the information is still accurate.

I'm going to guess there's a bug to do with symbol merging across files.

Reply:
Should we maybe also move this to a `markUsage` callback? I don't think this currently introduces any new bugs, since this is not called from anywhere else, and the future use I have in mind for this in the minimalist resolver will not really care about the usages.